prov/opx/src/fi_opx_init.c

```diff
@@ -0,0 +1,481 @@
+
+		/*
+		 * man/fi_endpoint.3
+		 *
+		 *   fi_tx_attr::caps
+		 *
+		 *   "... If the caps field is 0 on input to fi_getinfo(3), the
+		 *   caps value from the fi_info structure will be used."
+		 */
+		if (hints->tx_attr->caps) {
+			fi->tx_attr->caps = hints->tx_attr->caps;
+		}
+
+		/* adjust parameters down from what requested if required */
+		fi->tx_attr->op_flags = hints->tx_attr->op_flags;
+	} else if (hints && hints->caps) {
+		fi->tx_attr->caps = hints->caps;
+	}
+
+	memcpy(fi->rx_attr, fi_opx_global.default_rx_attr, sizeof(*fi->rx_attr));
+	if (hints && hints->rx_attr) {
+
+		/*
+		 * man/fi_endpoint.3
+		 *
+		 *   fi_rx_attr::caps
+		 *
+		 *   "... If the caps field is 0 on input to fi_getinfo(3), the
+		 *   caps value from the fi_info structure will be used."
+		 */
+		if (hints->rx_attr->caps) {
+			fi->rx_attr->caps = hints->rx_attr->caps;
+		}
+
+		/* adjust parameters down from what requested if required */
+		fi->rx_attr->op_flags = hints->rx_attr->op_flags;
+		if (hints->rx_attr->total_buffered_recv > 0 &&
+			hints->rx_attr->total_buffered_recv < fi_opx_global.default_rx_attr->total_buffered_recv)
+				fi->rx_attr->total_buffered_recv = hints->rx_attr->total_buffered_recv;
+	} else if (hints && hints->caps) {
+		fi->rx_attr->caps = hints->caps;
+	}
+
+	caps = fi->caps | fi->tx_attr->caps | fi->rx_attr->caps;
+
+	/*
+	 * man/fi_domain.3
+	 *
+	 * On input to fi_getinfo, a user may set this (fi_domain_attr::domain)
+	 * to an opened domain instance to restrict output to the given domain.
+	 * On output from fi_getinfo, if no domain was specified, but the user
+	 * has an opened instance of the named domain, this (fi_domain_attr::domain)
+	 * will reference the first opened instance. If no instance has been
+	 * opened, this field will be NULL.
+	 */
+
+	ret = fi_opx_choose_domain(caps, fi->domain_attr,
+		(hints)?(hints->domain_attr):NULL);
+	if (ret) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"cannot find appropriate domain\n");
+		goto err;
+	}
+
+	memcpy(fi->ep_attr, fi_opx_global.default_ep_attr, sizeof(*fi->ep_attr));
+	if (hints && hints->ep_attr) {
+		/* adjust parameters down from what requested if required */
+		fi->ep_attr->type	= hints->ep_attr->type;
+		if (hints->ep_attr->max_msg_size > 0 &&
+			hints->ep_attr->max_msg_size <= fi_opx_global.default_ep_attr->max_msg_size)
+				fi->ep_attr->max_msg_size = hints->ep_attr->max_msg_size;
+
+		if (0 != hints->ep_attr->tx_ctx_cnt && hints->ep_attr->tx_ctx_cnt <= fi->ep_attr->tx_ctx_cnt)
+			fi->ep_attr->tx_ctx_cnt = hints->ep_attr->tx_ctx_cnt;	/* TODO - check */
+
+		if (0 != hints->ep_attr->rx_ctx_cnt && hints->ep_attr->rx_ctx_cnt <= fi->ep_attr->rx_ctx_cnt)
+			fi->ep_attr->rx_ctx_cnt = hints->ep_attr->rx_ctx_cnt;	/* TODO - check */
+	}
+
+	return 0;
+err:
+	if (fi->domain_attr->name) free(fi->domain_attr->name);
+	if (fi->fabric_attr->name) free(fi->fabric_attr->name);
+	if (fi->fabric_attr->prov_name) free(fi->fabric_attr->prov_name);
```

Comment:
This frees the names, but leaves the pointers set.  If fi_freeinfo() is called to free the 'fi' structure because this function failed, it'll try to free the memory again.

Reply:
I will address this in the next push.